BUG: Fix serialization with Sphinx 7.3 (#1289)
sphinx-gallery/sphinx-gallery: doc/configuration.rst

dictionaries as input and returns a list of strings that are passed as
additional command line arguments to the interpreter.

> An example in ``doc/sphinxext.py`` (assuming you have added ``doc/`` to ``sys.path`` so
> that ``sphinxext.reset_argv`` can be resolved by importlib) could be::

Reviewer: ```suggestion
A ``reset_argv`` example defined in ``doc/sphinxext.py`` (assuming you have added ``doc/`` to ``sys.path`` so
that ``sphinxext.reset_argv`` can be resolved by importlib) could be::
```
nit

    from pathlib import Path

    def reset_argv(sphinx_gallery_conf, script_vars):
        src_file = Path(script_vars['src_file']).name
        if src_file == 'example1.py':
            return ['-a', '1']
        elif src_file == 'example2.py':
            return ['-a', '2']
        else:
            return []

which is included in the configuration dictionary as::

    sphinx_gallery_conf = {
        ...
        'reset_argv': "sphinxext.reset_argv",
    }

> which is then used by Sphinx-Gallery as::

Reviewer: I think the lines below this need to be updated.

:class:`NumberOfCodeLinesSortKey(src_dir) <sphinx_gallery.sorting.NumberOfCodeLinesSortKey>`,
which sorts the files based on the number of code lines::

    sphinx_gallery_conf = {
        ...
        'within_subsection_order': "NumberOfCodeLinesSortKey",
    }

Note that you must provide a *class* here, not an instance of the class. Because of how
sphinx caches the environment, this should be provided as a fully qualified string
that can be imported (though passing a class directly is supported for backward
compatibility). Built in sphinx-gallery classes, however, for convenience can be
specified using just the stem for example as ``"NumberOfLinesSortKey"``, which is
functionally equivalent to providing the fully qualified name
``"sphinx_gallery.sorting.NumberOfCodeLinesSortKey"``.

> In addition, multiple convenience classes are provided for use with
> ``within_subsection_order``:

Reviewer: ```suggestion
Built in convenience classes supported by ``within_subsection_order``:
```

:class:`NumberOfCodeLinesSortKey(src_dir) <sphinx_gallery.sorting.NumberOfCodeLinesSortKey>`,
which sorts the files based on the number of code lines::

    sphinx_gallery_conf = {
        ...
        'within_subsection_order': "NumberOfCodeLinesSortKey",
    }

> Note that you must provide a *class* here, not an instance of the class. Because of how

Reviewer: Stupid question, but if this is a string, does it matter if this is a class or instance of a class?

Author: Yes, it must be a class because we instantiate it internally. Hence the emphasis on it being a class not an instance here

:class:`NumberOfCodeLinesSortKey(src_dir) <sphinx_gallery.sorting.NumberOfCodeLinesSortKey>`,
which sorts the files based on the number of code lines::

    sphinx_gallery_conf = {
        ...

>         'within_subsection_order': "NumberOfCodeLinesSortKey",

Author: I'd rather highlight the shortcut version here. I think it's mostly what people will want to use. But I'll clarify the language that follows a bit.

    }

Note that you must provide a *class* here, not an instance of the class. Because of how

> sphinx caches the environment, this should be provided as a fully qualified string

Reviewer 2: That's not true. There was no behavior change in sphinx 7.3.0. there is just a warning now in case the config is not being cached.

* :ref:`subsection_order <sub_gallery_order>`
* :ref:`minigallery_sort_order <minigallery_order>`

The best way to do this is to make use of a single instance of each class you want
to use in some fully resolvable path, and then tell sphinx-gallery to resolve the
callable from the fully qualified name. For example, you could write a function
``def my_sorter`` and put it in ``mymod/utils.py``, then use::

    sphinx_gallery_conf = {
    #...,
    "minigallery_sort_order": "mymod.utils.my_sorter",
    #...
    }

Another option is to add documentation-specific stuff in a different path and ensure
that it can be resolved at build time. For example, you could create a file
``doc/sphinxext.py``::

    def plotted_sorter(fname):
        return not fname.startswith("plot_"), fname
    )

And set in your configuration::

    sys.path.insert(0, os.path.dirname(__file__))

    sphinx_gallery_conf = {
    #...,
    "minigallery_sort_order": "sphinxext.plotted_sorter",
    #...
    }

> And sphinx-gallery would resolve ``"sphinxext.function_sorter"`` to the
> ``function_sorter`` object because the ``doc/`` directory is first on the path.

Reviewer: ```suggestion
And sphinx-gallery would resolve ``"sphinxext.plotted_sorter"`` to the
``plotted_sorter`` object because the ``doc/`` directory is first on the path.
```

* :ref:`subsection_order <sub_gallery_order>`
* :ref:`minigallery_sort_order <minigallery_order>`

The best way to do this is to make use of a single instance of each class you want
to use in some fully resolvable path, and then tell sphinx-gallery to resolve the

> callable from the fully qualified name. For example, you could write a function

Reviewer: This does not seem to marry with the `CHANGES.rst` notes:

> 1. ``"mymod.utils.reset_argv"`` where you define a ``class _ResetArgv`` class in
   ``mymod.utils`` and instantiate it in your ``conf.py`` as
   ``reset_argv = _ResetArgv()``.

Author: Nice catch, fixed!
